### PR TITLE
fixes after code review

### PR DESCRIPTION
**Pull request by filippD**

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -5,49 +5,39 @@
...
+			get_symbol
+		elsif @player1 && @player1.symbol == player_symbol
+			puts "This symbol is already taken!"
+			get_symbol
 		else
-			puts 'Please make a valid move'
-			move(player)
+			return player_symbol
 		end
-	end
```
Comment (heyapricot) [inline lib/game.rb:29]: I like the idea of this. Just be careful with recursion (What happens when somebody enters the wrong symbol A LOT of times. You could fill the stack and program crashes ☠️)

File: lib/game.rb [hunk 2]
```diff
@@ -5,49 +5,39 @@
...
+		if player_symbol.length != 1
+			puts "Your symbol should be a single character!"
+			get_symbol
+		elsif @player1 && @player1.symbol == player_symbol
+			puts "This symbol is already taken!"
+			get_symbol
 		else
-			puts 'Please make a valid move'
-			move(player)
+			return player_symbol
```
Comment (heyapricot) [inline lib/game.rb:24]: You can omit the final else and just have the method return the player_symbol.
Also, ruby methods return the last called variable so you can omit the return statement and just type the variable name

File: lib/game.rb [hunk 3]
```diff
@@ -56,33 +46,62 @@ def continue
...
 	    choice = gets.chomp.upcase
-	    case
-		when choice == "N"
-			game_start
-		when choice == "C"
-			change_names
+	    case choice
+		when "G"
+			puts "Starting the game"
+			add_players if @player1 == nil && @player2 == nil
```
Comment (heyapricot) [inline lib/game.rb:85]: Is the conditional right? Do you want to add players only of BOTH players are missing? One player missing is enough reason to add the players again don't you think?

File: lib/game.rb [hunk 4]
```diff
@@ -56,33 +46,62 @@ def continue
...
+			add_board if @board == nil
```
Comment (heyapricot) [inline lib/game.rb:86]: Favor the use of `.nil?`

File: lib/game.rb [hunk 5]
```diff
@@ -110,27 +129,48 @@ def score
...
+				puts 'Invalid move'
+				move(player)
+			end
+		elsif player_input.length == 1 && player_input[0].upcase == 'Q'
+			@wanna_quit = true
+		else
+			puts 'Please make a valid move'
+			move(player)
+		end
+	end
```
Comment (heyapricot) [inline lib/game.rb:155]: This method does two things, validates input and then writes to the board.
The validation part is a bit messy, it relies on the board_valid input method but also on some conditionals and string substitution with a regular expression. It does all these alternating with writing a move to the board. 

Wouldn't it be better to have a single method that deals with the validation for `move` and once we are sure the move input is valid we pass it to the `move` method?

File: lib/game.rb [hunk 6]
```diff
@@ -110,27 +129,48 @@ def score
...
+		else
+			puts 'Please make a valid move'
+			move(player)
+		end
+	end	
+
 	def play
-		until player_won? || board_full?
+		current_player = @player1
+		until @board.player_won?(@player1.symbol) || @board.player_won?(@player2.symbol) || @board.full? || @wanna_quit
```
Comment (heyapricot) [inline lib/game.rb:159]: Do we need to check against a specific player? For the game to end we just need to check if there is a sequence of symbols in a row, a column or a diagonal. Once we know there is a sequence we can determine the winner based on the symbol or the current turn. 

Checking for a specific player symbol twice is not optimal

File: lib/board.rb [hunk 7]
```diff
@@ -1,33 +1,63 @@
...
+	end
+  
+  	def init_combin(dim)
+  		arr = []
+        arr  += Array.new(dim ){|l| Array.new(dim){|i| dim * l + i }}  
+        arr  += Array.new(dim){|l| Array.new(dim){|i|  l + (i*dim)}}
+        arr  += Array.new(1){|l| Array.new(dim){|i|  i+ (i*dim)}}
+        arr  += Array.new(1){|l| Array.new(dim){|i|  i*(dim-1) + (dim-1)}}
+        return arr                      
+    end
```
Comment (heyapricot) [inline lib/board.rb:16]: In general, avoid naming variables with only one character. Find the balance between being brief and explaining what a variable does/stores

File: lib/board.rb [hunk 8]
```diff
@@ -1,33 +1,63 @@
...
+
+    def valid_input?(arr)
+		if arr.length == 2 && 
+		(ALPHABET[0...@dim].any? {|val| val == arr[1].upcase}) &&
+		((0...@dim).to_a.any? {|element| element == arr[0].to_i})
+			return true
+		end
+		false
+	end
```
Comment (heyapricot) [inline lib/board.rb:25]: You never use this method inside the `board` object. You could validate every move before it is written or you could put the validator outside and assume you are getting a valid input. Either way is fine. Just pick one strategy, not half validation in one object and the other half on another. (If you assume you are getting a valid input that doesn't mean you shouldn't protect against possible errors like index out of reach).

File: lib/board.rb [hunk 9]
```diff
@@ -1,33 +1,63 @@
...
+
+	def empty_space?(i)
+		return true if @matrix[i] == nil
+		false
+	end
```
Comment (heyapricot) [inline lib/board.rb:30]: Exactly what I meant with 'half validation'. 
Choose one strategy:
- Validate the input on `Game` (outside the `board` object) and then pass the input to the board
- Validate input on `board` (Different methods for validation maybe?) and run the validators before a move is made. Then `Game` just has to call board.make_move (or similar). [I prefer this strategy but is up to you]

File: lib/board.rb [hunk 10]
```diff
@@ -1,33 +1,63 @@
...
 
-	def change_matrix(symbol, index)
-		@matrix[index] = symbol
+	def change_matrix(symbol, move_index)
+		@matrix[move_index] = symbol
 	end
 
 	def clear
-		@matrix.each {|element| element = nil}
+		@matrix = Array.new(dim**2)
```
Comment (heyapricot) [inline lib/board.rb:61]: You forgot the `@` before `dim`. Also, do you really need to call `Array.new(number)`? I mean, is it necessary that you have an array full of `nil` values?

**Review by heyapricot — CHANGES_REQUESTED** [review]
# Let's polish this gem

Hello! 👋🏼 Tic tac toe looks nice. I liked the menu you built. That said, I found some areas where we could improve performance and scalability. Your tic-tac-toe could end up being used by millions of users. Who knows? 

So, please address the comments I've left and after that send a new code review request. As always, feel free to reach over crisp for further discussion. Thank you!

File: lib/game.rb (hunk 3 above)
Comment (filippD) [reply lib/game.rb:85]: Well, the method adds 2 players so it doesn't really matter

Comment (filippD) [reply lib/game.rb:85]: Changed it to OR statement though

**Pull request merged**